c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+    /// The provided rewrite affects code that would be discarded by a rewrite of a containing
+    /// expression.  `.0` is the span of the containing expression, and `.1` is its rewrite.
+    _Discarded(S, Box<Rewrite<S>>),
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+struct RewriteTree<S = Span> {
+    span: S,
+    rw: Rewrite<S>,
+    /// Child nodes, representing rewrites that are contained entirely within the current rewrite.
+    /// Each child's `span` is contained within `self.span`, no child's `span` overlaps any other
+    /// child's `span`, and children are sorted by position (`span.lo()`).
+    children: Vec<RewriteTree<S>>,
+}
+
+trait SpanLike: Copy + Eq {
+    fn lo(self) -> BytePos;
+    fn hi(self) -> BytePos;
+    fn contains(self, other: Self) -> bool;
+    fn overlaps(self, other: Self) -> bool;
+}
+
+impl SpanLike for Span {
+    fn lo(self) -> BytePos {
+        self.lo()
+    }
+    fn hi(self) -> BytePos {
+        self.hi()
+    }
+    fn contains(self, other: Self) -> bool {
+        self.contains(other)
+    }
+    fn overlaps(self, other: Self) -> bool {
+        self.overlaps(other)
+    }
+}
+
+impl SpanLike for (u32, u32) {
+    fn lo(self) -> BytePos {
+        BytePos(self.0)
+    }
+    fn hi(self) -> BytePos {
+        BytePos(self.1)
+    }
+    fn contains(self, other: Self) -> bool {
+        self.0 <= other.0 && other.1 <= self.1
+    }
+    fn overlaps(self, other: Self) -> bool {
+        self.0 < other.1 && other.0 < self.1
+    }
+}
+
+impl<S: SpanLike> RewriteTree<S> {
+    pub fn build(
+        mut rws: Vec<(S, Rewrite<S>)>,
+    ) -> (Vec<RewriteTree<S>>, Vec<(S, Rewrite<S>, RewriteError<S>)>) {
+        // Sort by start position and then by decreasing length, so that each parent span comes
+        // before all its children.
+        rws.sort_by_key(|&(ref s, _)| (s.lo(), Reverse(s.hi() - s.lo())));
+
+        // The stack contains a `RewriteTree` node for the most recently processed item from `rws`
+        // along with all of its ancestors (nodes whose spans strictly contain its span).
+        // `stack[0]` is the outermost ancestor, and `stack.last()` is the most recently pushed
+        // node.  Nodes are "committed" once they are known to have no more children; when
+        // committed, the node is moved into the `children` list of its parent or (if it has no
+        // parent) into the output list `out`.
+        let mut stack = Vec::<RewriteTree<S>>::new();
+        let mut out = Vec::new();
+        let mut errs = Vec::new();
+
+        fn commit<S: SpanLike>(
+            stack: &mut Vec<RewriteTree<S>>,
+            out: &mut Vec<RewriteTree<S>>,
+            rt: RewriteTree<S>,
+        ) {
+            if let Some(parent) = stack.last_mut() {
+                debug_assert!(parent.span.contains(rt.span));
+                debug_assert!(parent
+                    .children
+                    .last()
+                    .map_or(true, |ch| ch.span.hi() <= rt.span.lo()),);
```

Comment:
why is this not `ch.span.hi() <= rt.span.hi()`? could you explain the semantic property you're looking to assert here?

Reply:
Added a comment.  `a.hi <= b.lo` checks if `b` comes after `a` without overlapping.